File: srt_editor.go
// srt_editor.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Subtitle struct {
	Index int
	Start time.Duration
	End   time.Duration
	Text  string
}

func parseTime(s string) (time.Duration, error) {
	s = strings.TrimSpace(s)
	s = strings.Replace(s, ",", ".", 1)
	parts := strings.Split(s, ":")
	if len(parts) != 3 {
		return 0, fmt.Errorf("invalid time format")
	}
	h, _ := strconv.Atoi(parts[0])
	m, _ := strconv.Atoi(parts[1])
	secParts := strings.Split(parts[2], ".")
	sec, _ := strconv.Atoi(secParts[0])
	ms := 0
	if len(secParts) > 1 {
		ms, _ = strconv.Atoi(secParts[1])
	}
	d := time.Duration(h)*time.Hour + time.Duration(m)*time.Minute + time.Duration(sec)*time.Second + time.Duration(ms)*time.Millisecond
	return d, nil
}

func formatTime(d time.Duration) string {
	h := int(d.Hours())
	m := int(d.Minutes()) % 60
	s := int(d.Seconds()) % 60
	ms := int(d.Milliseconds()) % 1000
	return fmt.Sprintf("%02d:%02d:%02d,%03d", h, m, s, ms)
}

func parseSRT(content string) []Subtitle {
	var subs []Subtitle
	blocks := strings.Split(strings.TrimSpace(content), "\n\n")
	for _, block := range blocks {
		lines := strings.Split(block, "\n")
		if len(lines) < 3 {
			continue
		}
		var idx int
		var start time.Duration
		var end time.Duration
		var textLines []string
		// первый может быть номером
		lineIdx := 0
		if idx, err := strconv.Atoi(lines[0]); err == nil {
			lineIdx = 1
			_ = idx // мы будем перенумеровывать
		}
		if lineIdx >= len(lines) {
			continue
		}
		timeLine := lines[lineIdx]
		parts := strings.Split(timeLine, "-->")
		if len(parts) != 2 {
			continue
		}
		start, err1 := parseTime(parts[0])
		end, err2 := parseTime(parts[1])
		if err1 != nil || err2 != nil {
			continue
		}
		text := strings.Join(lines[lineIdx+1:], "\n")
		sub := Subtitle{Index: 0, Start: start, End: end, Text: text}
		subs = append(subs, sub)
	}
	// перенумеровываем
	for i := range subs {
		subs[i].Index = i + 1
	}
	return subs
}

func formatSRT(subs []Subtitle) string {
	var lines []string
	for _, sub := range subs {
		lines = append(lines, strconv.Itoa(sub.Index))
		lines = append(lines, fmt.Sprintf("%s --> %s", formatTime(sub.Start), formatTime(sub.End)))
		lines = append(lines, sub.Text)
		lines = append(lines, "")
	}
	return strings.Join(lines, "\n")
}

type Editor struct {
	filename string
	subs     []Subtitle
}

func (e *Editor) load(filename string) error {
	data, err := os.ReadFile(filename)
	if err != nil {
		return err
	}
	e.subs = parseSRT(string(data))
	e.filename = filename
	fmt.Printf("Загружено %d субтитров.\n", len(e.subs))
	return nil
}

func (e *Editor) save(filename string) error {
	if filename == "" {
		filename = e.filename
	}
	content := formatSRT(e.subs)
	return os.WriteFile(filename, []byte(content), 0644)
}

func (e *Editor) shift(ms int) {
	delta := time.Duration(ms) * time.Millisecond
	for i := range e.subs {
		e.subs[i].Start += delta
		e.subs[i].End += delta
	}
	fmt.Printf("Сдвиг на %d мс выполнен.\n", ms)
}

func (e *Editor) renumber() {
	for i := range e.subs {
		e.subs[i].Index = i + 1
	}
	fmt.Println("Перенумерация выполнена.")
}

func (e *Editor) validate() []string {
	var errs []string
	if len(e.subs) == 0 {
		errs = append(errs, "Нет субтитров.")
	}
	var prevEnd time.Duration
	for i, sub := range e.subs {
		if sub.Start >= sub.End {
			errs = append(errs, fmt.Sprintf("#%d: начало >= конец", i+1))
		}
		if i > 0 && sub.Start < prevEnd {
			errs = append(errs, fmt.Sprintf("#%d: перекрытие с предыдущим", i+1))
		}
		prevEnd = sub.End
	}
	return errs
}

func (e *Editor) list() {
	for _, sub := range e.subs {
		fmt.Printf("#%d  %s --> %s\n%s\n\n", sub.Index, formatTime(sub.Start), formatTime(sub.End), sub.Text)
	}
}

func (e *Editor) add(startStr, endStr, text string) {
	start, err1 := parseTime(startStr)
	end, err2 := parseTime(endStr)
	if err1 != nil || err2 != nil {
		fmt.Println("Неверный формат времени.")
		return
	}
	sub := Subtitle{Index: len(e.subs) + 1, Start: start, End: end, Text: text}
	e.subs = append(e.subs, sub)
	e.renumber()
	fmt.Println("Субтитр добавлен.")
}

func (e *Editor) edit(idx int, startStr, endStr, text string) {
	if idx < 1 || idx > len(e.subs) {
		fmt.Println("Неверный номер.")
		return
	}
	sub := &e.subs[idx-1]
	if startStr != "" {
		if st, err := parseTime(startStr); err == nil {
			sub.Start = st
		}
	}
	if endStr != "" {
		if en, err := parseTime(endStr); err == nil {
			sub.End = en
		}
	}
	if text != "" {
		sub.Text = text
	}
	fmt.Println("Субтитр обновлён.")
}

func (e *Editor) delete(idx int) {
	if idx < 1 || idx > len(e.subs) {
		fmt.Println("Неверный номер.")
		return
	}
	e.subs = append(e.subs[:idx-1], e.subs[idx:]...)
	e.renumber()
	fmt.Println("Субтитр удалён.")
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Использование: go run srt_editor.go [--shift мс] [--check] [--output file] файл.srt")
		return
	}
	args := os.Args[1:]
	filename := ""
	shiftMs := 0
	checkOnly := false
	outputFile := ""
	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--shift":
			if i+1 < len(args) {
				shiftMs, _ = strconv.Atoi(args[i+1])
				i++
			}
		case "--check":
			checkOnly = true
		case "--output":
			if i+1 < len(args) {
				outputFile = args[i+1]
				i++
			}
		default:
			filename = args[i]
		}
	}
	if filename == "" {
		fmt.Println("Не указан файл.")
		return
	}
	editor := &Editor{}
	if err := editor.load(filename); err != nil {
		fmt.Println("Ошибка загрузки:", err)
		return
	}
	if checkOnly {
		errs := editor.validate()
		if len(errs) == 0 {
			fmt.Println("Субтитры корректны.")
		} else {
			fmt.Println("Ошибки:")
			for _, e := range errs {
				fmt.Println("  " + e)
			}
		}
		return
	}
	if shiftMs != 0 {
		editor.shift(shiftMs)
		if outputFile != "" {
			editor.save(outputFile)
		} else {
			editor.save("")
		}
		return
	}
	// interactive
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Редактор субтитров. Введите help для списка команд.")
	for {
		fmt.Print("> ")
		if !scanner.Scan() {
			break
		}
		cmd := strings.TrimSpace(scanner.Text())
		if cmd == "" {
			continue
		}
		parts := strings.Fields(cmd)
		op := parts[0]
		switch op {
		case "quit":
			return
		case "help":
			fmt.Println("list, show <номер>, add, edit <номер> [время_начала] [время_конца] [текст], delete <номер>, shift <мс>, renumber, validate, save, quit")
		case "list":
			editor.list()
		case "show":
			if len(parts) < 2 {
				fmt.Println("Укажите номер.")
				continue
			}
			idx, _ := strconv.Atoi(parts[1])
			if idx < 1 || idx > len(editor.subs) {
				fmt.Println("Неверный номер.")
			} else {
				sub := editor.subs[idx-1]
				fmt.Printf("#%d  %s --> %s\n%s\n", sub.Index, formatTime(sub.Start), formatTime(sub.End), sub.Text)
			}
		case "add":
			fmt.Print("Начало (HH:MM:SS,mmm): ")
			scanner.Scan()
			start := scanner.Text()
			fmt.Print("Конец (HH:MM:SS,mmm): ")
			scanner.Scan()
			end := scanner.Text()
			fmt.Print("Текст (введите END для завершения): ")
			var lines []string
			for {
				scanner.Scan()
				line := scanner.Text()
				if line == "END" {
					break
				}
				lines = append(lines, line)
			}
			text := strings.Join(lines, "\n")
			editor.add(start, end, text)
		case "edit":
			if len(parts) < 2 {
				fmt.Println("Укажите номер.")
				continue
			}
			idx, _ := strconv.Atoi(parts[1])
			fmt.Print("Новое начало (оставьте пустым): ")
			scanner.Scan()
			start := scanner.Text()
			fmt.Print("Новый конец (оставьте пустым): ")
			scanner.Scan()
			end := scanner.Text()
			fmt.Print("Новый текст (оставьте пустым): ")
			scanner.Scan()
			text := scanner.Text()
			editor.edit(idx, start, end, text)
		case "delete":
			if len(parts) < 2 {
				fmt.Println("Укажите номер.")
				continue
			}
			idx, _ := strconv.Atoi(parts[1])
			editor.delete(idx)
		case "shift":
			if len(parts) < 2 {
				fmt.Println("Укажите сдвиг в мс.")
				continue
			}
			ms, _ := strconv.Atoi(parts[1])
			editor.shift(ms)
		case "renumber":
			editor.renumber()
		case "validate":
			errs := editor.validate()
			if len(errs) == 0 {
				fmt.Println("OK")
			} else {
				fmt.Println("Ошибки:")
				for _, e := range errs {
					fmt.Println("  " + e)
				}
			}
		case "save":
			if len(parts) > 1 {
				editor.save(parts[1])
			} else {
				editor.save("")
			}
			fmt.Println("Сохранено.")
		default:
			fmt.Println("Неизвестная команда.")
		}
	}
}

File: srt_editor.rs
// srt_editor.rs
use std::fs;
use std::io::{self, Write, BufRead};
use std::time::Duration;
use std::str::FromStr;
use std::fmt;

#[derive(Debug, Clone)]
struct Subtitle {
    index: usize,
    start: Duration,
    end: Duration,
    text: String,
}

fn parse_time(s: &str) -> Result<Duration, String> {
    let s = s.trim().replace(',', ".");
    let parts: Vec<&str> = s.split(':').collect();
    if parts.len() != 3 {
        return Err("Invalid time format".to_string());
    }
    let h: u64 = parts[0].parse().map_err(|_| "Hours")?;
    let m: u64 = parts[1].parse().map_err(|_| "Minutes")?;
    let sec_parts: Vec<&str> = parts[2].split('.').collect();
    let sec: u64 = sec_parts[0].parse().map_err(|_| "Seconds")?;
    let ms: u64 = if sec_parts.len() > 1 {
        sec_parts[1].parse().unwrap_or(0)
    } else { 0 };
    Ok(Duration::from_secs(h*3600 + m*60 + sec) + Duration::from_millis(ms))
}

fn format_time(d: Duration) -> String {
    let total_sec = d.as_secs();
    let h = total_sec / 3600;
    let m = (total_sec % 3600) / 60;
    let s = total_sec % 60;
    let ms = d.subsec_millis();
    format!("{:02}:{:02}:{:02},{:03}", h, m, s, ms)
}

fn parse_srt(content: &str) -> Vec<Subtitle> {
    let mut subs = Vec::new();
    let blocks: Vec<&str> = content.trim().split("\n\n").collect();
    for block in blocks {
        let lines: Vec<&str> = block.lines().collect();
        if lines.len() < 3 { continue; }
        let mut idx = 0;
        let mut line_idx = 0;
        if let Ok(num) = lines[0].parse::<usize>() {
            idx = num;
            line_idx = 1;
        }
        if line_idx >= lines.len() { continue; }
        let time_line = lines[line_idx];
        let time_parts: Vec<&str> = time_line.split("-->").collect();
        if time_parts.len() != 2 { continue; }
        let start = parse_time(time_parts[0].trim()).unwrap_or(Duration::from_secs(0));
        let end = parse_time(time_parts[1].trim()).unwrap_or(Duration::from_secs(0));
        let text = lines[line_idx+1..].join("\n");
        subs.push(Subtitle { index: idx, start, end, text });
    }
    for (i, sub) in subs.iter_mut().enumerate() {
        sub.index = i + 1;
    }
    subs
}

fn format_srt(subs: &[Subtitle]) -> String {
    let mut out = String::new();
    for sub in subs {
        out.push_str(&format!("{}\n", sub.index));
        out.push_str(&format!("{} --> {}\n", format_time(sub.start), format_time(sub.end)));
        out.push_str(&sub.text);
        out.push_str("\n\n");
    }
    out
}

struct Editor {
    filename: String,
    subs: Vec<Subtitle>,
}

impl Editor {
    fn load(filename: &str) -> Result<Self, String> {
        let content = fs::read_to_string(filename).map_err(|e| e.to_string())?;
        let subs = parse_srt(&content);
        Ok(Editor { filename: filename.to_string(), subs })
    }
    fn save(&self, filename: Option<&str>) -> Result<(), String> {
        let fname = filename.unwrap_or(&self.filename);
        let content = format_srt(&self.subs);
        fs::write(fname, content).map_err(|e| e.to_string())
    }
    fn shift(&mut self, ms: i64) {
        let delta = if ms >= 0 {
            Duration::from_millis(ms as u64)
        } else {
            Duration::from_millis((-ms) as u64)
        };
        for sub in &mut self.subs {
            if ms >= 0 {
                sub.start += delta;
                sub.end += delta;
            } else {
                if sub.start >= delta { sub.start -= delta; }
                if sub.end >= delta { sub.end -= delta; }
            }
        }
        println!("Сдвиг на {} мс выполнен.", ms);
    }
    fn renumber(&mut self) {
        for (i, sub) in self.subs.iter_mut().enumerate() {
            sub.index = i + 1;
        }
        println!("Перенумерация выполнена.");
    }
    fn validate(&self) -> Vec<String> {
        let mut errs = Vec::new();
        if self.subs.is_empty() {
            errs.push("Нет субтитров.".to_string());
        }
        let mut prev_end = Duration::from_secs(0);
        for (i, sub) in self.subs.iter().enumerate() {
            if sub.start >= sub.end {
                errs.push(format!("#{}: начало >= конец", i+1));
            }
            if i > 0 && sub.start < prev_end {
                errs.push(format!("#{}: перекрытие с предыдущим", i+1));
            }
            prev_end = sub.end;
        }
        errs
    }
    fn list(&self) {
        for sub in &self.subs {
            println!("#{}  {} --> {}\n{}\n", sub.index, format_time(sub.start), format_time(sub.end), sub.text);
        }
    }
    fn add(&mut self, start_str: &str, end_str: &str, text: &str) {
        let start = parse_time(start_str).unwrap_or(Duration::from_secs(0));
        let end = parse_time(end_str).unwrap_or(Duration::from_secs(0));
        let sub = Subtitle { index: self.subs.len()+1, start, end, text: text.to_string() };
        self.subs.push(sub);
        self.renumber();
        println!("Субтитр добавлен.");
    }
    fn edit(&mut self, idx: usize, start_str: Option<&str>, end_str: Option<&str>, text: Option<&str>) {
        if idx < 1 || idx > self.subs.len() {
            println!("Неверный номер.");
            return;
        }
        let sub = &mut self.subs[idx-1];
        if let Some(s) = start_str {
            if let Ok(st) = parse_time(s) { sub.start = st; }
        }
        if let Some(e) = end_str {
            if let Ok(en) = parse_time(e) { sub.end = en; }
        }
        if let Some(t) = text {
            sub.text = t.to_string();
        }
        println!("Субтитр обновлён.");
    }
    fn delete(&mut self, idx: usize) {
        if idx < 1 || idx > self.subs.len() {
            println!("Неверный номер.");
            return;
        }
        self.subs.remove(idx-1);
        self.renumber();
        println!("Субтитр удалён.");
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        println!("Использование: cargo run -- [--shift мс] [--check] [--output file] файл.srt");
        return;
    }
    let mut filename = String::new();
    let mut shift_ms = 0;
    let mut check_only = false;
    let mut output_file = None;
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--shift" => {
                if i+1 < args.len() {
                    shift_ms = args[i+1].parse().unwrap_or(0);
                    i += 2;
                } else { i += 1; }
            }
            "--check" => { check_only = true; i += 1; }
            "--output" => {
                if i+1 < args.len() {
                    output_file = Some(args[i+1].clone());
                    i += 2;
                } else { i += 1; }
            }
            _ => { filename = args[i].clone(); i += 1; }
        }
    }
    if filename.is_empty() {
        println!("Не указан файл.");
        return;
    }
    let mut editor = match Editor::load(&filename) {
        Ok(e) => e,
        Err(e) => { println!("Ошибка загрузки: {}", e); return; }
    };
    if check_only {
        let errs = editor.validate();
        if errs.is_empty() {
            println!("Субтитры корректны.");
        } else {
            println!("Ошибки:");
            for e in errs { println!("  {}", e); }
        }
        return;
    }
    if shift_ms != 0 {
        editor.shift(shift_ms);
        if let Some(out) = output_file {
            editor.save(Some(&out)).unwrap();
        } else {
            editor.save(None).unwrap();
        }
        return;
    }

    let stdin = io::stdin();
    let mut stdout = io::stdout();
    println!("Редактор субтитров. Введите help для списка команд.");
    loop {
        print!("> ");
        stdout.flush().unwrap();
        let mut line = String::new();
        stdin.read_line(&mut line).unwrap();
        let cmd = line.trim();
        if cmd.is_empty() { continue; }
        let parts: Vec<&str> = cmd.split_whitespace().collect();
        let op = parts[0];
        match op {
            "quit" => break,
            "help" => println!("list, show <номер>, add, edit <номер> [время_начала] [время_конца] [текст], delete <номер>, shift <мс>, renumber, validate, save, quit"),
            "list" => editor.list(),
            "show" => {
                if parts.len() < 2 { println!("Укажите номер."); continue; }
                if let Ok(idx) = parts[1].parse::<usize>() {
                    if idx >= 1 && idx <= editor.subs.len() {
                        let sub = &editor.subs[idx-1];
                        println!("#{}  {} --> {}\n{}", sub.index, format_time(sub.start), format_time(sub.end), sub.text);
                    } else { println!("Неверный номер."); }
                }
            }
            "add" => {
                print!("Начало (HH:MM:SS,mmm): "); stdout.flush().unwrap();
                let mut start = String::new(); stdin.read_line(&mut start).unwrap();
                print!("Конец (HH:MM:SS,mmm): "); stdout.flush().unwrap();
                let mut end = String::new(); stdin.read_line(&mut end).unwrap();
                println!("Текст (введите END для завершения):");
                let mut text_lines = Vec::new();
                loop {
                    let mut line = String::new();
                    stdin.read_line(&mut line).unwrap();
                    if line.trim() == "END" { break; }
                    text_lines.push(line.trim().to_string());
                }
                let text = text_lines.join("\n");
                editor.add(&start.trim(), &end.trim(), &text);
            }
            "edit" => {
                if parts.len() < 2 { println!("Укажите номер."); continue; }
                let idx: usize = parts[1].parse().unwrap_or(0);
                print!("Новое начало (оставьте пустым): "); stdout.flush().unwrap();
                let mut start = String::new(); stdin.read_line(&mut start).unwrap();
                print!("Новый конец (оставьте пустым): "); stdout.flush().unwrap();
                let mut end = String::new(); stdin.read_line(&mut end).unwrap();
                print!("Новый текст (оставьте пустым): "); stdout.flush().unwrap();
                let mut text = String::new(); stdin.read_line(&mut text).unwrap();
                editor.edit(idx, if start.trim().is_empty() { None } else { Some(start.trim()) },
                            if end.trim().is_empty() { None } else { Some(end.trim()) },
                            if text.trim().is_empty() { None } else { Some(text.trim()) });
            }
            "delete" => {
                if parts.len() < 2 { println!("Укажите номер."); continue; }
                let idx: usize = parts[1].parse().unwrap_or(0);
                editor.delete(idx);
            }
            "shift" => {
                if parts.len() < 2 { println!("Укажите сдвиг в мс."); continue; }
                let ms: i64 = parts[1].parse().unwrap_or(0);
                editor.shift(ms);
            }
            "renumber" => editor.renumber(),
            "validate" => {
                let errs = editor.validate();
                if errs.is_empty() { println!("OK"); } else {
                    println!("Ошибки:");
                    for e in errs { println!("  {}", e); }
                }
            }
            "save" => {
                let fname = if parts.len() > 1 { Some(parts[1]) } else { None };
                if let Err(e) = editor.save(fname) {
                    println!("Ошибка сохранения: {}", e);
                } else {
                    println!("Сохранено.");
                }
            }
            _ => println!("Неизвестная команда."),
        }
    }
}
